API - If entity exist, use PATCH on create action

kuzmany opened the pull request:
…API controller

<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🔴🟢 <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | 🔴🟢
| Deprecations?                          | 🔴🟢
| BC breaks? (use the c.x branch)        | 🔴🟢
| Automated tests included?              | 🔴🟢 <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

This PR change method to process data in API for create hook.
Before If entity exist, it update it, but clear all others data what was not send in payload due this lines https://github.com/mautic/mautic/blob/3c1ee87dddb9a9da0ca0058f92bad4809fa876de/app/bundles/ApiBundle/Controller/CommonApiController.php#L493



<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create API call POST /api/contacts/new

```
{
    "email": "example@example.com",
    "firstname": "name"
}
```
3. Create API call POST /api/contacts/new

```
{
    "email": "example@example.com"
}
```
4. See like firstname was removed 
5. After PR should works properly

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

kuzmany committed:
refactor: Dynamically set HTTP method based on entity's existence in API controller

escopecz commented:
It seem to be more convenient the way you propose. I don't see any issue. Just add tests to get this merged. And target the 6.x branch. I'm not sure if it could break some current API workflows for anyone but it's safer to add it to a breaking backward compatibilty release.

kuzmany commented:
@escopecz Ok, will manage it.

driskell commented:
Just curious - doesn't make the create API behave exactly the same as the edit API with `PUT`. Just feels like this behaviour already exists on the edit `PUT`.

driskell commented:
Also I think if `overwriteWithBlank` is provided in the request to create it doesn't overwrite existing with blank if not provided in the request.

driskell commented:
Sorry deleted last 2 comments. Just checking how we use this (we hadn't seen any issues) and indeed it looks like only the bulk create endpoint does a `PATCH` if the entity already exists but the non-bulk version doesn't, so there's a disparity (we haven't used the non-bulk).

So perhaps workaround for now @kuzmany is to use `/api/contacts/bulk/new` instead of `/api/contacts/new` and this PR looks to be exactly the right fix.

kuzmany committed:
Merge branch 'refs/heads/6.x' into api-use-patch-during-create

kuzmany commented:
@escopecz currentAPI tests passed https://github.com/mautic/api-library/actions/runs/13197039503
Do you need another on on API side?

kuzmany committed:
docs: Update UPGRADE-6.0.md with API change in CommonApiController

kuzmany committed:
chore: Update environment to dev and add focus.html file

kuzmany committed:
chore: Update environment to production and remove unused focus.html file

escopecz commented:
@kuzmany yes, it would be really nice to have a test covering this endpoint that it behaves as you describe in the description. Just to make sure it will continue to work this way even after a few years.

kuzmany committed:
Merge remote-tracking branch 'origin/6.0' into api-use-patch-during-create

kuzmany committed:
Merge branch '6.0' into api-use-patch-during-create

escopecz commented:
@kuzmany do you think you could try t[his new AI command](https://github.com/mautic/mautic/pull/15102) for generating functional tests on this change? I think this deserves a test.

kuzmany committed:
Merge branch '6.0' into api-use-patch-during-create

kuzmany committed:
test: Add unit tests for CommonApiController PATCH/POST method selection

- Add tests to verify newEntityAction uses POST for new entities (no ID) and PATCH for existing entities (with ID)
- Implement data provider pattern for cleaner test structure
- Add helper methods to reduce code duplication in test setup
- All tests pass ensuring PR functionality works as expected

kuzmany committed:
Merge branch 'api-use-patch-during-create' of https://github.com/kuzmany/mautic into api-use-patch-during-create

kuzmany committed:
fix: Add proper type declarations to CommonApiControllerTest

- Add return types to all methods
- Add parameter types including nullable types
- Add proper array type annotations for data providers
- Fix method signature compatibility with parent class
- All PHPStan type issues resolved while maintaining test functionality

kuzmany committed:
fix: Add PHPDoc type annotation for processForm $entity parameter

- Add @param object $entity annotation to satisfy PHPStan
- Maintains compatibility with parent class method signature
- Resolves missingType.parameter PHPStan issue

kuzmany committed:
test: Add functional test for PATCH/POST method selection in newEntityAction

- Add test verifying POST method creates new contact (HTTP 201)
- Add test verifying existing contact with ID uses PATCH method (HTTP 200)
- Confirms the CommonApiController change works end-to-end via real API calls
- Tests actual behavior users will experience through the API

kuzmany committed:
revert: Remove all changes from CommonApiControllerTest.php

- Restore CommonApiControllerTest.php to original state
- Remove unit tests for PATCH/POST method selection
- Remove PHPStan type fixes and helper methods
- Keep functional test in LeadApiControllerFunctionalTest.php for end-to-end coverage

kuzmany commented:
@escopecz added

kuzmany committed:
Merge branch '6.0' into api-use-patch-during-create

escopecz reviewed (commented):
@kuzmany the test is passing even if I revert the fix. So do we need the code change if it behaves as you expect?
